core/services/blockheaderfeeder/block_header_feeder.go

```diff
@@ -0,0 +1,247 @@
+		err = f.batchBHS.StoreVerifyHeader(ctx, blockRange, blockHeaders)
+		if err != nil {
+			return errors.Wrap(err, "storing block headers")
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < uint64(fromBlock); block++ {
+			if _, ok := f.stored[block]; ok {
+				delete(f.stored, block)
+				f.lggr.Debugw("Pruned block from stored cache",
+					"block", block, "latestBlock", latestBlock)
+			}
+		}
+	}
+	f.lastRunBlock = latestBlock
+	return nil
+}
+
+func (f *BlockHeaderFeeder) findLowestBlockNumberWithoutBlockhash(ctx context.Context, lggr logger.Logger, blockToRequests map[uint64]map[string]struct{}) *big.Int {
+	var min *big.Int
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) == 0 {
+			continue
+		}
+		if _, ok := f.stored[block]; ok {
+			// Already stored
+			continue
+		}
+		stored, err := f.bhs.IsStored(ctx, block)
+		if err != nil {
+			lggr.Warnw("Failed to check if block is already stored",
+				"error", err,
+				"block", block)
+		} else if stored {
+			lggr.Infow("Blockhash already stored",
+				"block", block, "unfulfilledReqIDs", blockhashstore.LimitReqIDs(unfulfilledReqs))
+			f.stored[block] = struct{}{}
+			continue
+		}
+		blockNumber := big.NewInt(0).SetUint64(block)
+		if min == nil || min.Cmp(blockNumber) >= 0 {
+			min = blockNumber
+		}
+	}
+	return min
+}
+
+func (f *BlockHeaderFeeder) findEarliestBlockNumberWithBlockhash(ctx context.Context, lggr logger.Logger, startBlock, toBlock uint64) (*big.Int, error) {
+	from := startBlock
+	for i := startBlock; i < toBlock; i += uint64(f.getBlockhashesBatchSize) {
+		j := from + uint64(f.getBlockhashesBatchSize)
+		if j > toBlock {
+			j = toBlock
+		}
+
+		lggr.Debug(fmt.Sprintf("Looking for earliest block number with blockhash %v thru %v", i, j))
+
+		blockNumber := i
+		var blocks []*big.Int
+		for blockNumber < j {
+			blocks = append(blocks, big.NewInt(0).SetUint64(blockNumber))
+			blockNumber++
+		}
+
+		blockhashes, err := f.batchBHS.GetBlockhashes(ctx, blocks)
+		if err != nil {
+			return nil, errors.Wrap(err, "fetching blockhashes")
+		}
+
+		for idx, bh := range blockhashes {
+			if !bytes.Equal(bh[:], zeroHash[:]) {
+				earliestBlockNumber := i + uint64(idx)
+				lggr.Infow("found earliest block number with blockhash", "earliestBlockNumber", earliestBlockNumber, "blockhash", bh)
+				return big.NewInt(0).SetUint64(earliestBlockNumber), nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (f *BlockHeaderFeeder) getRlpHeadersBatch(ctx context.Context, blockRange []*big.Int) ([][]byte, error) {
+	var reqs []rpc.BatchElem
+	for _, num := range blockRange {
+		req := rpc.BatchElem{
+			Method: "eth_getHeaderByNumber",
+			// Get child block since it's the one that has the parent hash in its header.
+			Args:   []interface{}{num.Add(num, big.NewInt(1))},
+			Result: &types.Header{},
```

Comment:
I think you'll have to use our custom `evmtypes.Head` here instead of go-ethereum's `types.Header` to get always-correct blockhashes. Although I'm not sure if the RLP-encoding would work... 🤔

Reply:
synced offline. `types.Header` has an `EncodeRLP` method that forms the correct header. Avalanche will require using a different result type though.